# Project 3: Data Cleaning - Tidy up messy Datasets (Movies Dataset)

## First Steps

In [73]:
#Import libraries here / Max display = 30 
import pandas as pd
pd.options.display.max_columns = 30

In [74]:
#Load in dataset - Indenify what needs to be cleaned 
df = pd.read_csv("movies_metadata.csv", low_memory=False)

In [3]:
df

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45461  False                                                NaN         0   
45462  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                                                  genres  \
0      [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1      [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2      [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3      [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                         [{'id': 35, 'name': 'Comedy'}]   
...                                                  ...   
45461  [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...   
45462                      [{'id': 18, 'name': 'Drama'}]   
45463  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   
45464                                                 []   
45465                                                 []   

                                   homepage      id    imdb_id  \
0      http://toystory.disney.com/toy-story     862  tt0114709   
1                                       NaN    8844  tt0113497   
2                                       NaN   15602  tt0113228   
3                                       NaN   31357  tt0114885   
4                                       NaN   11862  tt0113041   
...                                     ...     ...        ...   
45461  http://www.imdb.com/title/tt6209470/  439050  tt6209470   
45462                                   NaN  111109  tt2028550   
45463                                   NaN   67758  tt0303758   
45464                                   NaN  227506  tt0008536   
45465                                   NaN  461257  tt6980792   

      original_language               original_title  \
0                    en                    Toy Story   
1                    en                      Jumanji   
2                    en             Grumpier Old Men   
3                    en            Waiting to Exhale   
4                    en  Father of the Bride Part II   
...                 ...                          ...   
45461                fa                      رگ خواب   
45462                tl          Siglo ng Pagluluwal   
45463                en                     Betrayal   
45464                en          Satana likuyushchiy   
45465                en                     Queerama   

                                                overview popularity  \
0      Led by Woody, Andy's toys live happily in his ...  21.946943   
1      When siblings Judy and Peter discover an encha...  17.015539   
2      A family wedding reignites the ancient feud be...    11.7129   
3      Cheated on, mistreated and stepped on, the wom...   3.859495   
4      Just when George Banks has recovered from his ...   8.387519   
...                                                  ...        ...   
45461        Rising and falling between a man and woman.   0.072051   
45462  An artist struggles to finish his work while a...   0.178241   
45463  When one of her hits goes wrong, a professiona...   0.903007   
45464  In a small town live two brothers, one a minis...   0.003503   
45465  50 years after decriminalisation of homosexual...   0.163015   

                            poster_path  \

In [75]:
##Summary of dataset - explore data types are correct/appropriate 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

In [76]:
#Explore nested columns to see format
df.genres[0]

"[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]"

In [77]:
df.belongs_to_collection[0]

"{'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}"

## Dropping irrelevant Columns


### Remove unnecessary columns for cleaner dataset

 - Data columns are removed based on what data needs to be extracted to fulfill a purpose. Depending on the end    result and what the data is needed will influence what columns should be dropped. 


In [7]:
#explore columns 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

In [78]:
df.adult.value_counts()

False                                                                                                                             45454
True                                                                                                                                  9
 - Written by Ørnås                                                                                                                   1
 Rune Balot goes to a casino connected to the October corporation to try to wrap up her case once and for all.                        1
 Avalanche Sharks tells the story of a bikini contest that turns into a horrifying affair when it is hit by a shark avalanche.        1
Name: adult, dtype: int64

In [79]:
# Drop columns not needed 
df.drop(columns = ['adult'], inplace = True)

In [80]:
df.drop(columns = ['imdb_id'], inplace = True)

In [81]:
df.drop(columns = ['original_title'], inplace = True)

In [82]:
df.drop(columns = ['video'], inplace= True)

In [83]:
df.drop(columns = ['homepage'], inplace= True)

## How to handle stringified JSON columns (Part 1)

In [84]:
import json
import ast

In [85]:
json_col = ["belongs_to_collection", "genres", "production_countries", 
            "production_companies", "spoken_languages"]

In [86]:
df.belongs_to_collection[0]

"{'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}"

In [87]:
#Convert stingfield JSON data into a list 
df.genres.apply(lambda x: json.loads(x.replace("'", '"')))[0]

[{'id': 16, 'name': 'Animation'},
 {'id': 35, 'name': 'Comedy'},
 {'id': 10751, 'name': 'Family'}]

In [88]:
df.genres.apply(ast.literal_eval)[0]

[{'id': 16, 'name': 'Animation'},
 {'id': 35, 'name': 'Comedy'},
 {'id': 10751, 'name': 'Family'}]

## How to handle stringified JSON columns (Part 2)

### Resolve the mixed data types in the column

In [89]:
import numpy as np

In [90]:
#Mixed data types in a column
df.belongs_to_collection

0        {'id': 10194, 'name': 'Toy Story Collection', ...
1                                                      NaN
2        {'id': 119050, 'name': 'Grumpy Old Men Collect...
3                                                      NaN
4        {'id': 96871, 'name': 'Father of the Bride Col...
                               ...                        
45461                                                  NaN
45462                                                  NaN
45463                                                  NaN
45464                                                  NaN
45465                                                  NaN
Name: belongs_to_collection, Length: 45466, dtype: object

In [91]:
# Create a Lambra function - isinstance to check If True needs to be cleaned/converted to missing values
df.belongs_to_collection.apply(lambda x: isinstance(x, str))

0         True
1        False
2         True
3        False
4         True
         ...  
45461    False
45462    False
45463    False
45464    False
45465    False
Name: belongs_to_collection, Length: 45466, dtype: bool

In [92]:
# If isinstance to check values are correct others converts to missing value 
df.belongs_to_collection = df.belongs_to_collection.apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else np.nan)

In [93]:
# Check first element
df.belongs_to_collection [0]

{'id': 10194,
 'name': 'Toy Story Collection',
 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg',
 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}

# Clean spoken languages & production companies  

In [94]:
df.spoken_languages

0                 [{'iso_639_1': 'en', 'name': 'English'}]
1        [{'iso_639_1': 'en', 'name': 'English'}, {'iso...
2                 [{'iso_639_1': 'en', 'name': 'English'}]
3                 [{'iso_639_1': 'en', 'name': 'English'}]
4                 [{'iso_639_1': 'en', 'name': 'English'}]
                               ...                        
45461               [{'iso_639_1': 'fa', 'name': 'فارسی'}]
45462                    [{'iso_639_1': 'tl', 'name': ''}]
45463             [{'iso_639_1': 'en', 'name': 'English'}]
45464                                                   []
45465             [{'iso_639_1': 'en', 'name': 'English'}]
Name: spoken_languages, Length: 45466, dtype: object

In [95]:
df.spoken_languages = df.spoken_languages.apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else np.nan)

In [96]:
df.production_countries

0        [{'iso_3166_1': 'US', 'name': 'United States o...
1        [{'iso_3166_1': 'US', 'name': 'United States o...
2        [{'iso_3166_1': 'US', 'name': 'United States o...
3        [{'iso_3166_1': 'US', 'name': 'United States o...
4        [{'iso_3166_1': 'US', 'name': 'United States o...
                               ...                        
45461               [{'iso_3166_1': 'IR', 'name': 'Iran'}]
45462        [{'iso_3166_1': 'PH', 'name': 'Philippines'}]
45463    [{'iso_3166_1': 'US', 'name': 'United States o...
45464             [{'iso_3166_1': 'RU', 'name': 'Russia'}]
45465     [{'iso_3166_1': 'GB', 'name': 'United Kingdom'}]
Name: production_countries, Length: 45466, dtype: object

In [27]:
df.production_countries = df.production_countries.apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else np.nan)

In [28]:
df.production_companies 

0           [{'name': 'Pixar Animation Studios', 'id': 3}]
1        [{'name': 'TriStar Pictures', 'id': 559}, {'na...
2        [{'name': 'Warner Bros.', 'id': 6194}, {'name'...
3        [{'name': 'Twentieth Century Fox Film Corporat...
4        [{'name': 'Sandollar Productions', 'id': 5842}...
                               ...                        
45461                                                   []
45462               [{'name': 'Sine Olivia', 'id': 19653}]
45463    [{'name': 'American World Pictures', 'id': 6165}]
45464                 [{'name': 'Yermoliev', 'id': 88753}]
45465                                                   []
Name: production_companies, Length: 45466, dtype: object

In [97]:
df.production_companies = df.production_companies.apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else np.nan)

In [98]:
#Explore dataset again - lists and dictionaries only 
df

belongs_to_collection    budget  \
0      {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1                                                    NaN  65000000   
2      {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3                                                    NaN  16000000   
4      {'id': 96871, 'name': 'Father of the Bride Col...         0   
...                                                  ...       ...   
45461                                                NaN         0   
45462                                                NaN         0   
45463                                                NaN         0   
45464                                                NaN         0   
45465                                                NaN         0   

                                                  genres      id  \
0      [{'id': 16, 'name': 'Animation'}, {'id': 35, '...     862   
1      [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...    8844   
2      [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   15602   
3      [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   31357   
4                         [{'id': 35, 'name': 'Comedy'}]   11862   
...                                                  ...     ...   
45461  [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...  439050   
45462                      [{'id': 18, 'name': 'Drama'}]  111109   
45463  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   67758   
45464                                                 []  227506   
45465                                                 []  461257   

      original_language                                           overview  \
0                    en  Led by Woody, Andy's toys live happily in his ...   
1                    en  When siblings Judy and Peter discover an encha...   
2                    en  A family wedding reignites the ancient feud be...   
3                    en  Cheated on, mistreated and stepped on, the wom...   
4                    en  Just when George Banks has recovered from his ...   
...                 ...                                                ...   
45461                fa        Rising and falling between a man and woman.   
45462                tl  An artist struggles to finish his work while a...   
45463                en  When one of her hits goes wrong, a professiona...   
45464                en  In a small town live two brothers, one a minis...   
45465                en  50 years after decriminalisation of homosexual...   

      popularity                       poster_path  \
0      21.946943  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1      17.015539  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2        11.7129  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   
3       3.859495  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg   
4       8.387519  /e64sOI48hQXyru7naBFyssKFxVd.jpg   
...          ...                               ...   
45461   0.072051  /jldsYflnId4tTWPx8es3uzsB1I8.jpg   
45462   0.178241  /xZkmxsNmYXJbKVsTRLLx3pqGHx7.jpg   
45463   0.903007  /d5bX92nDsISNhu3ZT69uHwmfCGw.jpg   
45464   0.003503  /aorBPO7ak8e8iJKT5OcqYxU3jlK.jpg   
45465   0.163015  /s5UkZt6NTsrS7ZF0Rh8nzupRlIU.jpg   

                                    production_companies  \
0         [{'name': 'Pixar Animation Studios', 'id': 3}]   
1      [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2      [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3      [{'name': 'Twentieth Century Fox Film Corporat...   
4      [{'name': 'Sandollar Productions', 'id': 5842}...   
...                                                  ...   
45461                                                 []   
45462             [{'name': 'Sine Olivia', 'id': 19653}]   
45463  [{'name': 'American World Pictures', 'id': 6165}]   
45464               [{'name': 'Yermoliev', 'id': 88753}]   
45465                                                 []   

                                    production_countries release_date 

## How to flatten nested Columns

In [99]:
#Explore first element - 4 key/value pairs  
df.belongs_to_collection[0]

{'id': 10194,
 'name': 'Toy Story Collection',
 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg',
 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}

In [100]:
#Check if dictionary or not - convert to missing values 
#Only extract name of movies
df.belongs_to_collection = df.belongs_to_collection.apply(lambda x: x['name'] if isinstance(x, dict) else np.nan) 

In [101]:
#Check value counts method 
df.belongs_to_collection.value_counts(dropna = False).head(20)

NaN                                       40975
The Bowery Boys                              29
Totò Collection                              27
James Bond Collection                        26
Zatôichi: The Blind Swordsman                26
The Carry On Collection                      25
Pokémon Collection                           22
Charlie Chan (Sidney Toler) Collection       21
Godzilla (Showa) Collection                  16
Charlie Chan (Warner Oland) Collection       15
Uuno Turhapuro                               15
Dragon Ball Z (Movie) Collection             15
The Land Before Time Collection              14
Monster High Collection                      14
George Carlin Comedy Collection              13
Sharpe Collection                            13
The Amityville Horror Collection             12
Johan Falk GSI Collection                    12
Sherlock Holmes (1939 series)                12
Friday the 13th Collection                   12
Name: belongs_to_collection, dtype: int6

In [102]:
#Put 3 genres into one string 
df.genres[0]

"[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]"

In [103]:
#Use pipe symbol and join together
df.genres = df.genres.apply(lambda x: "|".join(i['name'] for i in x))

TypeError: string indices must be integers

In [104]:
df.genres[0]

"[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]"

In [105]:
df.genres.value_counts(dropna = False).head(20)

[{'id': 18, 'name': 'Drama'}]                                                                    5000
[{'id': 35, 'name': 'Comedy'}]                                                                   3621
[{'id': 99, 'name': 'Documentary'}]                                                              2723
[]                                                                                               2442
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]                                  1301
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]                                      1135
[{'id': 27, 'name': 'Horror'}]                                                                    974
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]                                  930
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]     593
[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}]                       

In [106]:
#replace method for missing values 
df.genres.replace("", np.nan, inplace = True)

In [107]:
df.spoken_languages

0                 [{'iso_639_1': 'en', 'name': 'English'}]
1        [{'iso_639_1': 'en', 'name': 'English'}, {'iso...
2                 [{'iso_639_1': 'en', 'name': 'English'}]
3                 [{'iso_639_1': 'en', 'name': 'English'}]
4                 [{'iso_639_1': 'en', 'name': 'English'}]
                               ...                        
45461               [{'iso_639_1': 'fa', 'name': 'فارسی'}]
45462                    [{'iso_639_1': 'tl', 'name': ''}]
45463             [{'iso_639_1': 'en', 'name': 'English'}]
45464                                                   []
45465             [{'iso_639_1': 'en', 'name': 'English'}]
Name: spoken_languages, Length: 45466, dtype: object

In [108]:
# Convert missing values and join together with pipe symbol
df.spoken_languages = df.spoken_languages.apply(lambda x: "|".join(i['name'] for i in x) if isinstance(x, list) else np.nan) 

In [109]:
df.spoken_languages.value_counts(dropna = False).head(20)

English             22395
                     3952
Français             1853
日本語                  1289
Italiano             1218
Español               902
Pусский               807
Deutsch               762
English|Français      681
English|Español       572
हिन्दी                481
English|Deutsch       462
한국어/조선말               425
普通话                   347
English|Italiano      326
svenska               311
No Language           303
Português             275
suomi                 275
Polski                213
Name: spoken_languages, dtype: int64

 - 3952 missing values identified

In [112]:
#Replace method for missing values 
df.spoken_languages.replace("", np.nan, inplace = True)

In [113]:
# Explore prodction countries 
df.production_countries

0        [{'iso_3166_1': 'US', 'name': 'United States o...
1        [{'iso_3166_1': 'US', 'name': 'United States o...
2        [{'iso_3166_1': 'US', 'name': 'United States o...
3        [{'iso_3166_1': 'US', 'name': 'United States o...
4        [{'iso_3166_1': 'US', 'name': 'United States o...
                               ...                        
45461               [{'iso_3166_1': 'IR', 'name': 'Iran'}]
45462        [{'iso_3166_1': 'PH', 'name': 'Philippines'}]
45463    [{'iso_3166_1': 'US', 'name': 'United States o...
45464             [{'iso_3166_1': 'RU', 'name': 'Russia'}]
45465     [{'iso_3166_1': 'GB', 'name': 'United Kingdom'}]
Name: production_countries, Length: 45466, dtype: object

# Identified empty lists and values from production companies

In [114]:
# Convert missing values and join together with pipe symbol
df.production_countries = df.production_countries.apply(lambda x: "|".join(i['name'] for i in x) if isinstance(x, list) else np.nan)

In [115]:
#Replace missing values 
df.production_countries.value_counts(dropna = False).head(20)

NaN    45466
Name: production_countries, dtype: int64

In [116]:
df.production_countries.replace("", np.nan, inplace = True)

In [117]:
df.production_companies

0           [{'name': 'Pixar Animation Studios', 'id': 3}]
1        [{'name': 'TriStar Pictures', 'id': 559}, {'na...
2        [{'name': 'Warner Bros.', 'id': 6194}, {'name'...
3        [{'name': 'Twentieth Century Fox Film Corporat...
4        [{'name': 'Sandollar Productions', 'id': 5842}...
                               ...                        
45461                                                   []
45462               [{'name': 'Sine Olivia', 'id': 19653}]
45463    [{'name': 'American World Pictures', 'id': 6165}]
45464                 [{'name': 'Yermoliev', 'id': 88753}]
45465                                                   []
Name: production_companies, Length: 45466, dtype: object

In [118]:
df.production_companies = df.production_companies.apply(lambda x: "|".join(i['name'] for i in x) if isinstance(x, list) else np.nan) 

In [119]:
df.production_companies.value_counts(dropna = False).head(20)

                                          11875
Metro-Goldwyn-Mayer (MGM)                   742
Warner Bros.                                540
Paramount Pictures                          505
Twentieth Century Fox Film Corporation      439
Universal Pictures                          320
RKO Radio Pictures                          247
Columbia Pictures Corporation               207
Columbia Pictures                           146
Mosfilm                                     145
Walt Disney Pictures                         85
Universal International Pictures (UI)        82
New Line Cinema                              75
Walt Disney Productions                      75
Shaw Brothers                                71
Touchstone Pictures                          70
Toho Company                                 65
TriStar Pictures                             62
Orion Pictures                               61
Hammer Film Productions                      60
Name: production_companies, dtype: int64

In [120]:
df.production_companies.replace("", np.nan, inplace = True)

In [121]:
#Comapre first dataset to cleaned data set 
df.isna().sum()

belongs_to_collection    40975
budget                       0
genres                       0
id                           0
original_language           11
overview                   954
popularity                   5
poster_path                386
production_companies     11881
production_countries     45466
release_date                87
revenue                      6
runtime                    263
spoken_languages          3958
status                      87
tagline                  25054
title                        6
vote_average                 6
vote_count                   6
dtype: int64

In [122]:
pd.read_csv("movies_metadata.csv", low_memory=False).isna().sum()

adult                        0
belongs_to_collection    40972
budget                       0
genres                       0
homepage                 37684
id                           0
imdb_id                     17
original_language           11
original_title               0
overview                   954
popularity                   5
poster_path                386
production_companies         3
production_countries         3
release_date                87
revenue                      6
runtime                    263
spoken_languages             6
status                      87
tagline                  25054
title                        6
video                        6
vote_average                 6
vote_count                   6
dtype: int64

# Conclusion
- Filtered out missing values 
- removed irrelevent columns 

- Note - Having a lot of missing data does not mean necessarily mean we have unclean data. It could just be there is no data to explore. So we either remove the data or find some to replace it. 




## Cleaning Numerical Columns (Part 1)

In [125]:
#Summary of dataset - explore data types are correct/appropriate 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   belongs_to_collection  4491 non-null   object 
 1   budget                 45466 non-null  object 
 2   genres                 45466 non-null  object 
 3   id                     45466 non-null  object 
 4   original_language      45455 non-null  object 
 5   overview               44512 non-null  object 
 6   popularity             45461 non-null  object 
 7   poster_path            45080 non-null  object 
 8   production_companies   33585 non-null  object 
 9   production_countries   0 non-null      float64
 10  release_date           45379 non-null  object 
 11  revenue                45460 non-null  float64
 12  runtime                45203 non-null  float64
 13  spoken_languages       41508 non-null  object 
 14  status                 45379 non-null  object 
 15  ta

In [67]:
# Check if you can replce string to float if error then find another solution 
df.budget.astype("float")

AttributeError: 'DataFrame' object has no attribute 'budget'

In [68]:
# Cehck if you can replace string to float if error find another solution 
pd.to_numeric(df.budget)

AttributeError: 'DataFrame' object has no attribute 'budget'

In [126]:
# Try coerce - invalid parsing will be set to NaN as a solution to string error line 54 
df.budget = pd.to_numeric(df.budget, errors = "coerce")

In [127]:
df.budget.value_counts(dropna = False)

0.0           36573
5000000.0       286
10000000.0      259
20000000.0      243
2000000.0       242
              ...  
1119000.0         1
640000.0          1
17900000.0        1
4696772.0         1
2620000.0         1
Name: budget, Length: 1224, dtype: int64

 - 36, 573 missing data entries / 36,573 put as 0.0 as placeholder 

In [128]:
#Replace missing values 
df.budget = df.budget.replace(0, np.nan)

In [58]:
#Replace the budget values % by 1,000,000
df.budget = df.budget.div(1000000)

In [129]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   belongs_to_collection  4491 non-null   object 
 1   budget                 8890 non-null   float64
 2   genres                 45466 non-null  object 
 3   id                     45466 non-null  object 
 4   original_language      45455 non-null  object 
 5   overview               44512 non-null  object 
 6   popularity             45461 non-null  object 
 7   poster_path            45080 non-null  object 
 8   production_companies   33585 non-null  object 
 9   production_countries   0 non-null      float64
 10  release_date           45379 non-null  object 
 11  revenue                45460 non-null  float64
 12  runtime                45203 non-null  float64
 13  spoken_languages       41508 non-null  object 
 14  status                 45379 non-null  object 
 15  ta

 - Only 8,890 missing values now compared to previous 36,573

In [60]:
df.revenue.value_counts(dropna = False)

0.0            38052
12000000.0        20
10000000.0        19
11000000.0        19
2000000.0         18
               ...  
245724603.0        1
12212123.0         1
8203235.0          1
25900.0            1
12032983.0         1
Name: revenue, Length: 6864, dtype: int64

In [64]:
#Replace misisng values 
#Replace revenue in US Dollars 1,000,000
df.revenue = df.revenue.replace(0, np.nan)
df.revenue = df.revenue.div(1000000)

AttributeError: 'DataFrame' object has no attribute 'revenue'

In [62]:
df.rename(columns = {"revenue":"revenue_musd", "budget":"budget_musd"}, inplace = True)

In [124]:
#Summary of dataset - explore data types are correct/appropriate 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   belongs_to_collection  4491 non-null   object 
 1   budget                 45466 non-null  object 
 2   genres                 45466 non-null  object 
 3   id                     45466 non-null  object 
 4   original_language      45455 non-null  object 
 5   overview               44512 non-null  object 
 6   popularity             45461 non-null  object 
 7   poster_path            45080 non-null  object 
 8   production_companies   33585 non-null  object 
 9   production_countries   0 non-null      float64
 10  release_date           45379 non-null  object 
 11  revenue                45460 non-null  float64
 12  runtime                45203 non-null  float64
 13  spoken_languages       41508 non-null  object 
 14  status                 45379 non-null  object 
 15  ta

 - Have fewer missing columns than before 

## Cleaning Numerical Columns (Part 2)

In [130]:
df.runtime.value_counts(dropna = False).head(20)

90.0     2556
0.0      1558
100.0    1470
95.0     1412
93.0     1214
96.0     1104
92.0     1080
94.0     1062
91.0     1057
88.0     1032
97.0     1027
85.0     1024
98.0     1019
105.0    1002
89.0      958
87.0      919
110.0     850
86.0      846
99.0      794
102.0     791
Name: runtime, dtype: int64

 - Identified 0.0 for 1558 counts in dataset / missing values 

In [131]:
#Replace missing values with NaN
df.runtime = df.runtime.replace(0, np.nan)

In [132]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   belongs_to_collection  4491 non-null   object 
 1   budget                 8890 non-null   float64
 2   genres                 45466 non-null  object 
 3   id                     45466 non-null  object 
 4   original_language      45455 non-null  object 
 5   overview               44512 non-null  object 
 6   popularity             45461 non-null  object 
 7   poster_path            45080 non-null  object 
 8   production_companies   33585 non-null  object 
 9   production_countries   0 non-null      float64
 10  release_date           45379 non-null  object 
 11  revenue                45460 non-null  float64
 12  runtime                43645 non-null  float64
 13  spoken_languages       41508 non-null  object 
 14  status                 45379 non-null  object 
 15  ta

 - ID column should be numeric column try to convert it 

In [134]:
#try convert column to numeric if error use coerce method
pd.to_numeric(df.id)

ValueError: Unable to parse string "1997-08-20" at position 19730

In [135]:
#coerce method replaces missing values as NaN
df.id = pd.to_numeric(df.id, errors = "coerce")

In [136]:
df.id.value_counts(dropna = False).head(20)

141971.0    3
NaN         3
12600.0     2
18440.0     2
42495.0     2
152795.0    2
132641.0    2
14788.0     2
265189.0    2
4912.0      2
84198.0     2
298721.0    2
109962.0    2
5511.0      2
110428.0    2
22649.0     2
15028.0     2
10991.0     2
105045.0    2
13209.0     2
Name: id, dtype: int64

- Id columns should be a primary key and not have duplicates 

In [137]:
#Summary of dataset - explore data types are correct/appropriate 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   belongs_to_collection  4491 non-null   object 
 1   budget                 8890 non-null   float64
 2   genres                 45466 non-null  object 
 3   id                     45463 non-null  float64
 4   original_language      45455 non-null  object 
 5   overview               44512 non-null  object 
 6   popularity             45461 non-null  object 
 7   poster_path            45080 non-null  object 
 8   production_companies   33585 non-null  object 
 9   production_countries   0 non-null      float64
 10  release_date           45379 non-null  object 
 11  revenue                45460 non-null  float64
 12  runtime                43645 non-null  float64
 13  spoken_languages       41508 non-null  object 
 14  status                 45379 non-null  object 
 15  ta

 - Popularity should be numeric type not object - convert to numeric  

In [139]:
#Explore popularity dataset 
df.popularity

0        21.946943
1        17.015539
2          11.7129
3         3.859495
4         8.387519
           ...    
45461     0.072051
45462     0.178241
45463     0.903007
45464     0.003503
45465     0.163015
Name: popularity, Length: 45466, dtype: object

 - Data type should be float but is object

In [140]:
#Try convert to numeric if error then use coerce method
pd.to_numeric(df.popularity)

ValueError: Unable to parse string "Beware Of Frost Bites" at position 35587

In [143]:
#Create missing value for "Beware Of Frost Bites"
df.popularity = pd.to_numeric(df.popularity, errors='coerce')

In [144]:
df.popularity.value_counts(dropna = False).head(20)

0.000000    66
0.000001    56
0.000308    43
0.000220    40
0.000578    38
0.000844    38
0.001177    38
0.002001    28
0.003013    21
0.001393    19
0.003530    19
0.036471    18
0.002353    18
0.000603    16
0.001586    15
0.004425    14
0.000431    13
0.001021    13
0.004706    12
0.003553    11
Name: popularity, dtype: int64

 - 66 counts as 0.0 / could be posisble as 66 movies received 0 votes 

In [145]:
df.vote_count.value_counts(dropna = False).head(20)

1.0     3264
2.0     3132
0.0     2899
3.0     2787
4.0     2480
5.0     2097
6.0     1747
7.0     1570
8.0     1359
9.0     1194
10.0    1171
11.0     944
12.0     859
13.0     733
14.0     700
15.0     674
16.0     601
17.0     554
18.0     497
20.0     463
Name: vote_count, dtype: int64

In [146]:
df.vote_average.value_counts(dropna = False).head(20)

0.0    2998
6.0    2468
5.0    2001
7.0    1886
6.5    1722
6.3    1603
5.5    1381
5.8    1369
6.4    1350
6.7    1342
6.8    1324
6.1    1281
6.6    1263
6.2    1253
5.9    1196
5.3    1082
5.7    1046
6.9    1037
5.6    1006
7.3    1000
Name: vote_average, dtype: int64

 -  2,998 Movies with no rating  = 0.0 / misrepresents a fair rating for those movies 

In [147]:
# Filter all 0 counts for vote average
df.loc[df.vote_count == 0, "vote_average"]

83       0.0
107      0.0
126      0.0
132      0.0
137      0.0
        ... 
45432    0.0
45434    0.0
45452    0.0
45464    0.0
45465    0.0
Name: vote_average, Length: 2899, dtype: float64

In [148]:
#Change 0 to missing values 
df.loc[df.vote_count == 0, "vote_average"] = np.nan

In [149]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   belongs_to_collection  4491 non-null   object 
 1   budget                 8890 non-null   float64
 2   genres                 45466 non-null  object 
 3   id                     45463 non-null  float64
 4   original_language      45455 non-null  object 
 5   overview               44512 non-null  object 
 6   popularity             45460 non-null  float64
 7   poster_path            45080 non-null  object 
 8   production_companies   33585 non-null  object 
 9   production_countries   0 non-null      float64
 10  release_date           45379 non-null  object 
 11  revenue                45460 non-null  float64
 12  runtime                43645 non-null  float64
 13  spoken_languages       41508 non-null  object 
 14  status                 45379 non-null  object 
 15  ta

 - All numeric columns have been changed to the correct data type
 - replace missing values to NaN

## Cleaning DateTime Columns

In [ ]:
df.info()

In [ ]:
df.release_date

In [ ]:
#pd.to_datetime(df.release_date)

In [ ]:
df.release_date = pd.to_datetime(df.release_date, errors = "coerce")

In [ ]:
df.release_date.value_counts(dropna = False).head(20)

## Cleaning Text / String Columns

In [ ]:
df.info()

In [ ]:
df.original_language.value_counts(dropna = False).head(50)

In [ ]:
df.title

In [ ]:
df.title.value_counts(dropna = False).head(20)

In [ ]:
df.overview[0]

In [ ]:
df.overview.value_counts(dropna = False).head(20)

In [ ]:
df.overview.replace("No overview found.", np.nan, inplace = True)

In [ ]:
df.overview.replace("No Overview", np.nan, inplace = True)

In [ ]:
df.overview.replace("No movie overview available.", np.nan, inplace = True)

In [ ]:
df.overview.replace(" ", np.nan, inplace = True)

In [ ]:
df.overview.replace("No overview yet.", np.nan, inplace = True)

In [ ]:
df.tagline.value_counts(dropna = False).head(50)

In [ ]:
df.tagline.replace("-", np.nan, inplace = True)

## Removing Duplicates

In [ ]:
df[df.duplicated(keep =  False)].sort_values(by = "id")

In [ ]:
df.drop_duplicates(inplace = True)

In [ ]:
df[df.duplicated(subset = "id", keep = False)].sort_values(by = "id")

In [ ]:
df.drop_duplicates(subset = "id", inplace = True)

In [ ]:
df.id.value_counts(dropna = False)

## Handling Missing Values & Removing Observations

In [ ]:
df.info()

In [ ]:
df.isna().sum()

In [ ]:
df[df.title.isna()]

In [ ]:
df.dropna(subset = ["id", "title"], inplace = True)

In [ ]:
df.id = df.id.astype("int")

In [ ]:
df.notna().sum(axis = 1).value_counts().sort_values(ascending = False)

In [ ]:
df[df.notna().sum(axis = 1) == 6]

In [ ]:
df.dropna(thresh = 10, inplace = True)

In [ ]:
df.info()

In [ ]:
df.isna().sum()

## Final (Cleaning) Steps

In [ ]:
df.status.value_counts()

In [ ]:
df = df.loc[df.status == "Released"].copy()

In [ ]:
df.drop(columns = ["status"], inplace = True)

In [ ]:
col = ["id", "title", "tagline", "release_date", "genres", "belongs_to_collection", 
       "original_language", "budget_musd", "revenue_musd", "production_companies",
       "production_countries", "vote_count", "vote_average", "popularity", "runtime",
       "overview", "spoken_languages", "poster_path"]

In [ ]:
df = df.loc[:, col]

In [ ]:
df

In [ ]:
df.reset_index(drop = True, inplace =True)

In [ ]:
df.info()

In [ ]:
df.poster_path[0]

In [ ]:
base_poster_url = 'http://image.tmdb.org/t/p/w185/'
df.poster_path = "<img src='" + base_poster_url + df.poster_path + "' style='height:100px;'>"

In [ ]:
df.to_csv("movies_clean.csv", index = False)

In [ ]:
pd.read_csv("movies_clean.csv").info()